File: shared/ios/Keybase/Pusher.m
//
//  Pusher.m
//  Keybase
//
//  Created by Michael Maxim on 5/7/18.
//  Copyright © 2018 Keybase. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "Pusher.h"
@import UserNotifications;

@implementation PushNotifier {
}
- (void)localNotification:(NSString*)ident msg:(NSString*)msg badgeCount:(long)badgeCount soundName:(NSString*)soundName convID:(NSString*)convID typ:(NSString*)typ {
  UNMutableNotificationContent *localNotification = [UNMutableNotificationContent new];
  localNotification.sound = [UNNotificationSound soundNamed:soundName];
  NSNumber* badge = NULL;
  if (badgeCount >= 0) {
    badge = [NSNumber numberWithLong:badgeCount];
  }
  localNotification.badge = badge;
  localNotification.body = msg;
  localNotification.userInfo = @{ @"convID" : convID, @"type" : typ};
  UNNotificationRequest *request = [UNNotificationRequest requestWithIdentifier:ident content:localNotification trigger:nil];
  UNUserNotificationCenter* center = [UNUserNotificationCenter currentNotificationCenter];
  [center addNotificationRequest:request withCompletionHandler:^(NSError * _Nullable error) {
    if (error != nil) {
      NSLog(@"local notification failed: %@",error);
    }
  }];
}

- (void)displayChatNotification:(KeybaseChatNotification *)notification {
  NSString* ident = [NSString stringWithFormat:@"%@:%ld", notification.convID, notification.message.id_];
  NSString* msg;
  if (notification.isPlaintext && [notification.message.plaintext length] != 0) {
    if([ notification.message.from.keybaseUsername isEqualToString: notification.conversationName ] || [notification.conversationName length] == 0) {
      msg = [NSString stringWithFormat:@"%@: %@", notification.message.from.keybaseUsername, notification.message.plaintext];
    } else {
      msg = [NSString stringWithFormat:@"%@ (%@): %@", notification.message.from.keybaseUsername, notification.conversationName, notification.message.plaintext];
    }
  } else {
    msg = notification.message.serverMessage;
  }

  [self localNotification:ident msg:msg badgeCount:notification.badgeCount soundName:notification.soundName convID:notification.convID typ:@"chat.newmessage"];
}

@end

File: shared/ios/Keybase/Pusher.swift
import Foundation
import UserNotifications
import Keybasego
import os

private let log = Logger(subsystem: "com.keybase.app", category: "push")

// Serial so notifications post in the order Go scheduled them.
private let pushQueue = DispatchQueue(label: "com.keybase.app.push", qos: .userInitiated)

class PushNotifier: NSObject, Keybasego.KeybasePushNotifierProtocol {
  func localNotification(
    _ ident: String?, title: String?, msg: String?, badgeCount: Int, soundName: String?,
    convID: String?, typ: String?, uid: String?
  ) {
    // Invoked by Go over the gomobile cgo bridge, so this runs on a Go-runtime-managed thread,
    // not a real NSThread/dispatch queue. Hop to a GCD queue before touching UserNotifications.
    pushQueue.async {
      let content = UNMutableNotificationContent()
      if let soundName = soundName {
        content.sound = UNNotificationSound(named: UNNotificationSoundName(rawValue: soundName))
      }
      content.badge = (badgeCount >= 0) ? NSNumber(value: badgeCount) : nil
      content.title = title ?? ""
      content.body = msg ?? ""
      content.userInfo = ["convID": convID ?? "", "type": typ ?? "", "uid": uid ?? ""]
      let request = UNNotificationRequest(
        identifier: ident ?? UUID().uuidString, content: content, trigger: nil)
      UNUserNotificationCenter.current().add(request) { error in
        if let error = error {
          log.error("local notification failed: \(error.localizedDescription, privacy: .public)")
        }
      }
    }
  }

  // If we lost the race against the server's fallback timeout, the generic
  // "you have a new message" push for this message is already in Notification
  // Center. The local plaintext notification supersedes it, so remove it.
  private func removeDeliveredGenericNotification(convID: String, msgID: Int) {
    let center = UNUserNotificationCenter.current()
    center.getDeliveredNotifications { delivered in
      let ids = delivered.filter { note in
        let info = note.request.content.userInfo
        guard let type = info["type"] as? String, type == "chat.newmessage" else { return false }
        guard let c = info["convID"] as? String, c == convID else { return false }
        let m = (info["msgID"] as? NSNumber)?.intValue ?? Int(info["msgID"] as? String ?? "")
        return m == msgID
      }.map { $0.request.identifier }
      if !ids.isEmpty {
        center.removeDeliveredNotifications(withIdentifiers: ids)
      }
    }
  }

  func display(_ n: KeybaseChatNotification?) {
    guard let notification = n else { return }
    guard let message = notification.message else { return }

    let ident = "\(notification.convID):\(message.id_)"
    let msg: String
    if notification.isPlaintext && !message.plaintext.isEmpty {
      let username = message.from?.keybaseUsername ?? ""
      let convName = notification.conversationName
      msg = (username == convName || convName.isEmpty)
        ? "\(username): \(message.plaintext)"
        : "\(username) (\(convName)): \(message.plaintext)"
    } else {
      msg = message.serverMessage
    }
    let title = notification.title

    let soundName: String? = notification.soundName.isEmpty ? nil : notification.soundName
    let uid: String? = notification.uid.isEmpty ? nil : notification.uid
    localNotification(
      ident, title: title, msg: msg,
      badgeCount: notification.badgeCount,
      soundName: soundName,
      convID: notification.convID,
      typ: "chat.newmessage",
      uid: uid
    )
    removeDeliveredGenericNotification(convID: notification.convID, msgID: Int(message.id_))
  }
}
